invoker/core/src/test/java/com/google/cloud/functions/invoker/GcfEventsTest.java

```diff
@@ -87,6 +90,118 @@ public void checkAllProperties() throws IOException {
     assertThat(cloudEvent.getDataSchema()).isNull();
   }
 
+  // The next set of tests checks the result of using Gson to deserialize the JSON "data" field of the
+  // CloudEvent that we get from converting a legacy event. For the most part we're not testing much here,
+  // since the "data" field is essentially copied from the input legacy event. In some cases we adjust it,
+  // though.
+  // Later, when we have support for handling these types properly in Java, we can change the tests to use
+  // that. See https://github.com/googleapis/google-cloudevents-java
+
+  @Test
+  public void storageData() throws IOException {
+    Event legacyEvent = legacyEventForResource("storage.json");
+    CloudEvent cloudEvent = GcfEvents.convertToCloudEvent(legacyEvent);
+    Map<String, Object> data = cloudEventDataJson(cloudEvent);
+    assertThat(data).containsAtLeast(
+        "bucket", "some-bucket",
+        "timeCreated", "2020-04-23T07:38:57.230Z",
+        "generation", "1587627537231057",
+        "metageneration", "1",
+        "size", "352");
+  }
+
+  @Test
+  public void firestoreSimpleData() throws IOException {
+    Event legacyEvent = legacyEventForResource("firestore_simple.json");
+    CloudEvent cloudEvent = GcfEvents.convertToCloudEvent(legacyEvent);
+    Map<String, Object> data = cloudEventDataJson(cloudEvent);
+    Map<String, Object> expectedValue = Map.of(
+        "name", "projects/project-id/databases/(default)/documents/gcf-test/2Vm2mI1d0wIaK2Waj5to",
+        "createTime", "2020-04-23T09:58:53.211035Z",
+        "updateTime", "2020-04-23T12:00:27.247187Z",
+        "fields", Map.of(
+            "another test", Map.of("stringValue", "asd"),
+            "count", Map.of("integerValue", "4"),
+            "foo", Map.of("stringValue", "bar")));
+    Map<String, Object> expectedOldValue = Map.of(
+        "name", "projects/project-id/databases/(default)/documents/gcf-test/2Vm2mI1d0wIaK2Waj5to",
+        "createTime", "2020-04-23T09:58:53.211035Z",
+        "updateTime", "2020-04-23T12:00:27.247187Z",
+        "fields", Map.of(
+            "another test", Map.of("stringValue", "asd"),
+            "count", Map.of("integerValue", "3"),
+            "foo", Map.of("stringValue", "bar")));
+    assertThat(data).containsAtLeast(
+        "value", expectedValue,
+        "oldValue", expectedOldValue,
+        "updateMask", Map.of("fieldPaths", List.of("count")));
+  }
+
+  @Test
+  public void firestoreComplexData() throws IOException {
```

Comment:
This test in .NET was mostly because we have helpers to make Firestore easier to work with. You probably don't need this test here.

Reply:
Right, the comment on lines 93–98 says we're not testing much here, but I think it is useful to show that we can successfully dissect a complex value, even if it is currently quite tedious and manual.